python/ql/src/experimental/semmle/python/frameworks/SqlAlchemy.qll

```diff
@@ -0,0 +1,97 @@
+/**
+ * Provides classes modeling security-relevant aspects of the 'SqlAlchemy' package.
+ * See https://pypi.org/project/SQLAlchemy/.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.dataflow.new.TaintTracking
+private import semmle.python.ApiGraphs
+private import semmle.python.Concepts
+
+private module SqlAlchemy {
+  /**
+   * An instantization of a SqlAlchemy Session object.
+   * See https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session and
+   * https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker
+   */
+  private class SqlAlchemySessionInstance extends API::Node {
+    SqlAlchemySessionInstance() {
+      this in [
+          API::moduleImport("sqlalchemy.orm").getMember("Session").getReturn(),
+          API::moduleImport("sqlalchemy.orm").getMember("sessionmaker").getReturn().getReturn()
+        ]
+    }
+
+    override string toString() { result = "Use of SqlAlchemy Session instantization" }
+  }
+
+  /**
+   * An instantization of a SqlAlchemy Engine object.
+   * See https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine
+   */
+  private class SqlAlchemyEngineInstance extends API::Node {
+    SqlAlchemyEngineInstance() {
+      this = API::moduleImport("sqlalchemy").getMember("create_engine").getReturn()
+    }
+
+    override string toString() { result = "Use of SqlAlchemy create_engine member" }
+  }
+
+  /**
+   * An instantization of a SqlAlchemy Query object.
+   * See https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=query#sqlalchemy.orm.Query
+   */
+  private class SqlAlchemyQueryInstance extends API::Node {
+    SqlAlchemyQueryInstance() {
+      this = any(SqlAlchemySessionInstance sessionInstance).getMember("query").getReturn()
+    }
+
+    override string toString() { result = "Use of SqlAlchemy Session Query member" }
+  }
+
+  /**
+   * A call to `execute` meant to execute an SQL expression
+   * See the following links:
+   *   - https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=execute#sqlalchemy.engine.Connection.execute
+   *   - https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=execute#sqlalchemy.engine.Engine.execute
+   *   - https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=execute#sqlalchemy.orm.Session.execute
+   */
+  private class SqlAlchemyExecuteCall extends DataFlow::CallCfgNode, SqlExecution::Range {
+    SqlAlchemyExecuteCall() {
+      exists(SqlAlchemySessionInstance sessionInstance, SqlAlchemyEngineInstance engineInstance |
+        this = sessionInstance.getMember("execute").getACall() or
+        this = engineInstance.getMember("connect").getReturn().getMember("execute").getACall() or
+        this = engineInstance.getMember("begin").getReturn().getMember("execute").getACall()
+      )
```

Comment:
Notice that this code effective says:

There must exist a `SqlAlchemySessionInstance` and a `SqlAlchemyEngineInstance`, **and only then** will there be any members of the `SqlAlchemyExecuteCall` class.

I don't think that's what you intended to write, so I think instead this should be

```suggestion
    SqlAlchemyExecuteCall() {
      exists(SqlAlchemySessionInstance sessionInstance |
        this = sessionInstance.getMember("execute").getACall()
      )
      or
      exists(SqlAlchemyEngineInstance engineInstance |
        this = engineInstance.getMember("connect").getReturn().getMember("execute").getACall()
        or
        this = engineInstance.getMember("begin").getReturn().getMember("execute").getACall()
      )
```

if you rewrite things to use predicates instead of classes (as discussed above), I also think this can end up being a more straightforward predicate.

Lastly, I want to highlight a neat little trick. Instead of 

```codeql
        this = engineInstance.getMember("connect").getReturn().getMember("execute").getACall()
        or
        this = engineInstance.getMember("begin").getReturn().getMember("execute").getACall()
```

you can write

```codeql
        this = engineInstance.getMember(["connect", "begin"]).getReturn().getMember("execute").getACall()
```

Reply:
Thank you. I did not notice that illogical statement. I will be more careful next time.